rfcs/2020-08-31-3645-graphql-api.md

```diff
@@ -0,0 +1,279 @@
+
+* A type system that accommodates interfaces, unions, enums, primitives and custom scalar types. async-graphql provides built-in abstractions for chrono `DateTime` types, uuid, and other popular crates.
+
+* A known schema for errors. Snafu compatibility for `FieldResult<T>` custom errors.
+
+### Later -- Authorization
+
+Initial Vector observability will be single-instance, and available to anyone that has access to the configured port.
+
+Locking down access will initially rely on network configuration.
+
+Later, as we move into multi-instance observability and more granular API permissions, the requirement for user auth and persistence will surface.
+
+While those concerns are out-of-scope for this RFC, choosing a protocol that facilitates authentication is important to avoid backing ourselves into a corner.
+
+GraphQL is not opinionated with auth. We have any authorization mechanism available to us at the intersection of HTTP and WebSockets.
+
+In the previous Timber project, we appended an `Authorization: Bearer <jwt>` header for queries/mutations.
+
+For subscription, we passed a JWT along with the initial WebSocket connection payload; browsers pass limited headers with `Upgrade` requests, so this provided a neat approach to sidestep the lack of a comparable header for WebSockets. The JWT persisted for the life of the open WS.
+
+I anticipate doing something similar with the Vector API. async-graphql has a [Context](https://async-graphql.github.io/async-graphql/en/context.html) struct, which is typically used for passing in shared resources such as a database connection pool, or request-specific data such as the current user session.
+
+This is largely TBD, but the basic mechanisms are there to allow for flexible auth when we need it.
+
+## Doc-level Proposal
+
+A Vector observability layer has already been agreed internally. Work is underway.
+
+This proposal discusses the protocol which will govern communications between a running Vector instance, and a web UI and CLI.
+
+The result of this proposal won't directly impact user interaction with observability tooling.
+
+## Rationale
+
+We have built a decent body of experience with GraphQL at Timber, albeit on non-Vector projects.
+
+The tooling I am proposing here represents the same stack (save for swapping Apollo for Urql.)
+
+In a previous Timber project, we used [gqlgen](https://github.com/99designs/gqlgen) as our server library, written in Go. The tooling with async-graphql is similar, though Rust's language features enable a more composible 'code first' approach using macros.
+
+I believe the type of data we are consuming benefits from a strongly typed interface, and that compile-time client generation will significantly reduce the time-to-market of what will be already be a very complex front-end web app.
```

Comment:
Could you expand on the workflow here? If the interface types are generated at compile time from the rust code in vector, how do we then consume that in a strongly typed way from the UI? Is there a compilation output from Vector that we need to keep in sync with vector-ui? Do we lose type safety if the UI is on a different version from vector itself? Are there patterns for making changes without breaking compatibility?

I'm sure these are all well understood for people with GraphQL experience, but it'd be nice to lay the answers out a bit more explicitly for those of us who haven't used it.

Reply:
I've added a section titled ["Development Workflow", here.](https://github.com/timberio/vector/blob/093543af0243ee2c4c58b9da1b1e545bbfa4a08d/rfcs/2020-08-31-3645-graphql-api.md#development-workflow) (this points to a commit to permalink it; future viewers may wish to view this on master in case of changes.)

I've included a 16 minute live coding session showing how the UI interacts with the API. This is intended to augment the example code blocks already in the RFC, which I felt were easy to overlook and weren't very indicative of the benefits the front-end team inherits opting for GraphQL. The only way to appreciate it, IMO, is to write code querying live data, so hopefully this video makes it clearer where the potential speed-ups are vs. opting for an untyped / DIY approach.

Reply:
> Is there a compilation output from Vector that we need to keep in sync with vector-ui?

@sghall and I talked about this a while back, and there was a brief discussion in the internal pre-RFC with some ideas, but I don't think we had enough context back then (and maybe still don't) to reach a firm consesus.

The ideas were roughly:

1. Embedding the `index.html` equivalent in Vector with some dynamic values injected to set-up the target endpoint, and then loading the rest via a `<script>` tag pointing to a canonical URL. This would free us up from having to embed a (potentially quite weighty) set of JS/CSS files into Vector, and would simplify the story around local hosting. The downside is the requirement for an active Internet connection for the end-user (i.e. the person loading the admin UI, and _not_ Vector itself-- which could still be locked down.)

2. Embed everything in Vector, so the UI and Vector are always versioned, and always in sync. The downside is the extra 2-5mb added to the binary, which will likely keep growing for a while.

3. Embed nothing (or just index.html) and then have a separate `--ui-path` to point to the UI files. Pros: No drag on the binary. Cons: Difficult versioning/upgrade story; more complex config.

Would be interested to get some feedback on the above.

> Do we lose type safety if the UI is on a different version from vector itself? Are there patterns for making changes without breaking compatibility?

Yes. This is more insidious than just loosing type safety-- queries will error if fields or types have changed. There are various best practices that have been kicked around / are quasi-established for handling versioning (@jszwedko is probably the most clued up, here-- he worked through a guide relatively recently which tackled some of the more advanced concerns that can crop up), but the general take-away is to try to avoid breaking changes where they're not necessary, and that the client should be in sync. A similar concern domain as NoSQL (although the analogy breaks down with mutations, which would also error if the query didn't match.)

I consider this a good thing, because - like protobufs - we're guarding against undefined behavior in the type system before even attempting to parse a query. But it does mean the UI has to match a given version of Vector (or at least a given version of its schema.)

Reply:
Yeah,  coordinating the server and client pieces is a little tricky.  The UI is pulling the types from a running instance of the GraphQL server and so the process needs some thought.  Seems like we're going to need a CI process that will type check the UI code vs the schema to get 100% assurance (?).  Something like pulls the UI by tag (?) from GitHub (?) and checks the types and builds it.  As far as what exactly it builds and where it goes, like to get some guidance on that.   Big question obviously on whether we can assume an internet connection.  IMHO that will drive usage down if they need to go out to the web from a secure environment.

Reply:
>  Seems like we're going to need a CI process that will type check the UI code vs the schema to get 100% assurance

Yeah, I think that kinda thing is going to be the right move. I'd treat Vector as the compilation target. Whether it's embedding in the binary or some other co-location strategy, the UI should be bundled against a specific version of Vector and tested via the same `yarn workspace vortex graphql`-type live introspection in GH Actions. 

If it's remote, we could have `<script src="https://cdn.vector.dev/js/main.<vector_version>.js">` to keep them in sync. If it's bundled, then some CI approach that builds from [vector-ui](https://github.com/timberio/vector-ui).

cc @Hoverbear, @lukesteensen - do you have any initial thoughts on how you foresee static UI files integrating with Vector? The loading strategy and how we handle cross-repo compilation should be driven by those conclusions, IMO.

Reply:
> I consider this a good thing, because - like protobufs - we're guarding against undefined behavior in the type system before even attempting to parse a query. But it does mean the UI has to match a given version of Vector (or at least a given version of its schema.)

This is actually one of the benefits of protocol buffers, in that there are a specific set of rules for how you can evolve their definition that allows clients on different versions to continue to interoperate. For example, adding new fields is always compatible and older clients will simply ignore fields they don't know about.

Since it sounds like GraphQL is pretty much just JSON, I'd imagine the specific behavior there could depend on the implementation. I just want to be realistic and understand that UI and backend versions won't always be exactly the same, so the way we handle mismatches is going to matter. It could be as simple as forcing a refresh if we detect an incompatible change, but it'd be nice if there were some known safe migration process for GraphQL types.





> do you have any initial thoughts on how you foresee static UI files integrating with Vector?

I think we'd take the static files as an input to our release process and bundle them with our distribution artifacts in some fashion. Vector itself would then act as a static file server for them. I don't want to have to assume any connectivity other than to the vector process itself.

Reply:
> Since it sounds like GraphQL is pretty much just JSON, I'd imagine the specific behavior there could depend on the implementation.

A GraphQL query will always be valid, assuming the requested fields exist, and substituted input parameters match the expected type.

This is by design; one of the main intentions of GraphQL is to prevent over-fetch, which allows a query to request a subset of fields and not the full available set. Given that GraphQL data can be highly nested (the 'graph' in graphql), it wouldn't be practical or even possible to request every field, anyway -- it supports infinite recursion.

For example, a query could look like this:

```gql
query SourcesToSinks($beginsWith: String!){
  sources {
    id
    name
    transforms(filter: { name: { beginsWith: $beginsWith}}) {
      id
      name
      sinks {
        id
        name
      }
    }
  }
}
```

Note we don't specify the types. We just expect on the client, given prior introspection, that they will match expectations. Reasons this query could fail include:

- One or more of the fields don't exist
- The 'transforms' filter input isn't valid
- The nested `beginsWith` param isn't of type `String!`
- The client requesting doesn't have permission to request a given field
- A nesting/recursion/complexity limit has been defined in the server, which the query violates

Otherwise, it will pass.

This would also pass:

```gql
query JustSources {
  sources {
    id
    name
  }
}
```

Note in this case, we don't requested the related `transforms` (or sinks) at all. By design, we can request just what we need for a given query.

The practical effect to this is backwards compatibility for any added fields. If the query was written when `transforms` didn't exist, it would still work. Queries are validated at runtime by virtue of not receiving an error.

This is the same for all GraphQL clients, regardless of language or implementation.

A breaking change would be if we requested, say, `name`, and name didn't exist... or if `beginsWith` was modified to a type other than `String!` (in this case.)

Reply:
> I think we'd take the static files as an input to our release process and bundle them with our distribution artifacts in some fashion. Vector itself would then act as a static file server for them. I don't want to have to assume any connectivity other than to the vector process itself.

👍 sounds good. Do you foresee these being embedded in the binary itself? Or saved in a separate dir, which `vector` could be configured to point to? I don't have context re: existing distribution artifacts, so not sure if there's already a concept of third-party file deps we could leverage.